StreamWriter.cpp

```diff
@@ -1,11 +1,24 @@
 #include "StreamWriter.h"
+#include "XFile.h"
 #include <stdexcept>
 
 StreamWriter::~StreamWriter() { }
 
-FileStreamWriter::FileStreamWriter(const std::string& filename)
+FileStreamWriter::FileStreamWriter(const std::string& filename, unsigned int writeSettings)
 {
-	fileStream.open(filename.c_str(), std::ios::trunc | std::ios::out | std::ios::binary);
+	if (!FlagSet(writeSettings, FileWriteFlag::Overwrite)) {
```

Comment:
When you pre-check conditions before actually performing the operation, you run the risk of race conditions. There may be other threads and other processes concurrently accessing and modifying the filesystem. What you're checking may not be what you later operate on. There have actually been security vulnerabilities that resulted from such race conditions. It would be more advisable to translate the set of API flags to a set of system flags, and rely on the in-built checks of the underlying system. The underlying system API should be able to atomically check permissions while opening the file.

Reply:
Hmm, good point. That does seem like a rather limited selection. I guess I expected more options to be available under the hood.